### DISCUSSION TITLE
FreeBSD support?

### DISCUSSION THREAD

**Discussion by windymelt** [xremap/xremap#623] [Ideas]
Hi, thank you for nice software. I'm using this in other my Linux environments.

I tried to build xremap on FreeBSD 14.1 laptop but it raised error:

```
   Compiling xremap v0.10.0 (/home/windymelt/src/github.com/k0kubun/xremap)
error[E0432]: unresolved import `nix::sys::inotify`
  --> src/main.rs:13:15
   |
13 | use nix::sys::inotify::{AddWatchFlags, Inotify, InotifyEvent};
   |               ^^^^^^^ could not find `inotify` in `sys`

error[E0432]: unresolved import `nix::sys::timerfd`
  --> src/main.rs:16:15
   |
16 | use nix::sys::timerfd::{ClockId, TimerFd, TimerFlags};
   |               ^^^^^^^ could not find `timerfd` in `sys`

error[E0432]: unresolved import `nix::sys::inotify`
  --> src/config/mod.rs:20:15
   |
20 | use nix::sys::inotify::{AddWatchFlags, InitFlags, Inotify};
   |               ^^^^^^^ could not find `inotify` in `sys`

error[E0432]: unresolved import `nix::sys::inotify`
 --> src/device.rs:8:15
  |
8 | use nix::sys::inotify::{AddWatchFlags, InitFlags, Inotify};
  |               ^^^^^^^ could not find `inotify` in `sys`

error[E0432]: unresolved import `nix::sys::timerfd`
  --> src/event_handler.rs:16:15
   |
16 | use nix::sys::timerfd::{Expiration, TimerFd, TimerSetTimeFlags};
   |               ^^^^^^^ could not find `timerfd` in `sys`

For more information about this error, try `rustc --explain E0432`.
error: could not compile `xremap` (bin "xremap") due to 5 previous errors
```

Can we solve this? Should we port something?

FreeBSD have `kqueue`, and there seems like to be compat layer. https://www.freshports.org/devel/libinotify/

**Comment by k0kubun** [comment]
I'm open to merging contributions if it does work once we fix it. For example, switching the import and the code based on the OS is acceptable.

However, as a Linux user, I don't have an environment to develop it myself. If we add FreeBSD support, I'd like the person who added the support to be a maintainer for the platform.

**Comment by hpccc53** [comment]
Experimental FreeBSD support is added in `v0.15.6`. It has some limitations, that is [discribed here](https://github.com/xremap/xremap/blob/master/doc/reference_freebsd.md). @windymelt if you have feedback on whether it works, it would be appreciated.
